### ISSUE TITLE
SQL optimization on course loop

### ISSUE THREAD

**Issue by Mte90** [gocodebox/lifterlms#486]
Hi it's me again with other possible improvements from SQL side to do on lifter.

Looking on the Course page (the page with all the courses) there are queries like that:

```SELECT SQL_CALC_FOUND_ROWS wp_posts.ID
FROM wp_posts
INNER JOIN wp_postmeta
ON ( wp_posts.ID = wp_postmeta.post_id )
WHERE 1=1
AND ( wp_posts.ID NOT IN (
SELECT object_id
FROM wp_term_relationships
WHERE term_taxonomy_id IN (253) ) )
AND ( ( wp_postmeta.meta_key = '_llms_product_id'
AND wp_postmeta.meta_value = '19338' ) )
AND wp_posts.post_type = 'llms_access_plan'
AND (wp_posts.post_status = 'publish'
OR wp_posts.post_status = 'closed'
OR wp_posts.post_status = 'llms-completed'
OR wp_posts.post_status = 'llms-active'
OR wp_posts.post_status = 'llms-expired'
OR wp_posts.post_status = 'llms-on-hold'
OR wp_posts.post_status = 'llms-pending'
OR wp_posts.post_status = 'llms-cancelled'
OR wp_posts.post_status = 'llms-refunded'
OR wp_posts.post_status = 'llms-failed'
OR wp_posts.post_status = 'llms-txn-failed'
OR wp_posts.post_status = 'llms-txn-pending'
OR wp_posts.post_status = 'llms-txn-refunded'
OR wp_posts.post_status = 'llms-txn-succeeded'
OR wp_posts.post_status = 'private'
OR wp_posts.post_status = 'hidden')
GROUP BY wp_posts.ID
ORDER BY wp_posts.menu_order ASC
LIMIT 0, 10```

or

```SELECT SQL_CALC_FOUND_ROWS wp_posts.ID
FROM wp_posts
INNER JOIN wp_postmeta
ON ( wp_posts.ID = wp_postmeta.post_id )
WHERE 1=1
AND ( wp_posts.ID NOT IN (
SELECT object_id
FROM wp_term_relationships
WHERE term_taxonomy_id IN (253) ) )
AND ( ( wp_postmeta.meta_key = '_llms_product_id'
AND wp_postmeta.meta_value = '38' ) )
AND wp_posts.post_type = 'llms_access_plan'
AND (wp_posts.post_status = 'publish'
OR wp_posts.post_status = 'closed'
OR wp_posts.post_status = 'llms-completed'
OR wp_posts.post_status = 'llms-active'
OR wp_posts.post_status = 'llms-expired'
OR wp_posts.post_status = 'llms-on-hold'
OR wp_posts.post_status = 'llms-pending'
OR wp_posts.post_status = 'llms-cancelled'
OR wp_posts.post_status = 'llms-refunded'
OR wp_posts.post_status = 'llms-failed'
OR wp_posts.post_status = 'llms-txn-failed'
OR wp_posts.post_status = 'llms-txn-pending'
OR wp_posts.post_status = 'llms-txn-refunded'
OR wp_posts.post_status = 'llms-txn-succeeded'
OR wp_posts.post_status = 'private'
OR wp_posts.post_status = 'hidden')
GROUP BY wp_posts.ID
ORDER BY wp_posts.menu_order ASC
LIMIT 0, 10```

As you can see the only difference is the `meta_value` so maybe is better to do a single query with all the meta_value and later filter them by code to avoid 10 query (I have 10 courses).
Something like do wp_query that optimize the calls on loop and later organize the informations.

**Comment by Mte90** [comment]
As I can see that optimization will work also inside the courses.

**Comment by Mte90** [comment]
Yss sorry I was talking about the course catalog and the course page on the second comment.

**Comment by Mte90** [comment]
In any case I am talking about page like:

* Course page: https://skillsandmore.org/corso/crea-i-nuovi-layout-con-flexbox/
* Course catalog: https://skillsandmore.org/corsi/

**Comment by Mte90** [comment]
Any updates for this?

**Comment by Mte90** [comment]
I think that the code involved is https://github.com/gocodebox/lifterlms/blob/7fb73308111d78764c3934dc8792d3ecdc812ac5/includes/models/model.llms.product.php#L38
That is executed in different point for every product like on https://github.com/gocodebox/lifterlms/blob/c97b956ee0552dd22be9c89c5e3d6c803e6a7f60/includes/models/model.llms.course.php#L568 or https://github.com/gocodebox/lifterlms/blob/c13a3a9862b7039c096b4d557c6c8b906102890c/templates/product/pricing-table.php#L26

Probably the best way is:

* On a page where there are more then one product
* Do a query to get all the access plans of all of them
* Store that value somewhere

Maybe an idea to do that can be:

* Use object cache to cache all the access plans (and execute that on the course page as example)
* Inside the method that do the query use the cache and get only the value requested

In that way seems the best way with less code changed. Caching that will be not bad consdiering that is used everywhere from logged to non-logged users. Cache for every single product can be expensive and one unique can improve a lot.

**Comment by Mte90** [comment]
Or another solution can be on get_access_plans add a parameter like `$all` that change the `WP_Query` to not use `meta_value`.

* there will be all the plans
* cache this output
* from this output get the plan requested

This solution seems more easy then the other one but maybe cah change the behaviour of the class that is for single product and for all of them so maybe is better a feedback from you @thomasplevy for the best way to do it.

**Comment by thomasplevy** [comment]
@Mte90 thanks for your suggestions here... however, upon closer inspection I believe these queries are resulting from customizations you've made to your catalog template, specifically the HTML you've added to output prices on the course tiles.

By default, LifterLMS _does not_ output prices (as I'm sure you're aware). Our course prices come from access plans. So adding custom HTML via actions or customized templates is what's adding these calls to the postmeta table.

Yes, all of this is very much not ideal given your templates and parameters but this isn't an issue that I feel warrants a fix in the core due to the fact that it is coming from custom code.

I think that there's a compromise here:

We can add a filter to allow you to short circuit the call to 

https://github.com/gocodebox/lifterlms/blob/7fb73308111d78764c3934dc8792d3ecdc812ac5/includes/models/model.llms.product.php#L38

Once there's a filter here you could implement a custom function on the catalog page's `wp` action or `init` which does a single query to retrieve all the necessary access plan ids in whatever method you see fit. Store that result in the cache or as a global and then utilize the new filter to pull the data from the cache/global and return it early instead of executing another WP_Query as the function normally does.

Can you confirm if this _seems_ to make sense as a possible solution?

If so, I'll add a filter into the next release for you and consider this resolved when that's deployed.

**Comment by Mte90** [comment]
You are right our designer added a custom template of https://github.com/gocodebox/lifterlms/blob/master/templates/loop/content.php, called `content-course.php`.
This file contain:
```
$pricing_plans = $product->get_access_plans();
foreach ( $pricing_plans as $i => $p ){
		$price = $p->get_price( 'price' );
	}
if( isset($price) ) :
		?>
			<div class="price"><?php echo $price; ?></div>
		<?php endif; ?>```
Looking at the code get a shortcircuit for us will be a problem because we are using a template to that.
Probably when the loop is generated this kind of information inside the product class need to already available during the loop.

**Comment by thomasplevy** [comment]
@Mte90 

> Looking at the code get a shortcircuit for us will be a problem because we are using a template to that.
> Probably when the loop is generated this kind of information inside the product class need to already available during the loop.

If I write code that gathers all this data prior to the loop that would mean that I'd add a query/queries to the core codebase that are completely unnecessary for any site who hasn't customized their loop to include this pricing information. I'd optimize your site but I'd slow down all the other sites using LifterLMS.

If the solution I suggested won't work for you so you could add some custom code around your existing custom code than I'm not sure what else I can do for you here...

**Comment by Mte90** [comment]
You are right seems a lot something very specifically so there it is my proposal.

Add an option in the settings to add in the course page the price, in that way you can do the loop with this value and cache as you want and of course change also the template of the archive page to add this value.

In that way everyone that want to show the price (that is something reasonable after all we are selling them) has an option and not create issues with the others that doesn't care.
